File: Dev/C++/Mascaret/src/VEHA/Behavior/Activity/ActivityNode.cpp
#include "VEHA/Behavior/Activity/ActivityNode.h"
#include "VEHA/Behavior/Activity/ActionNode.h"
#include "VEHA/Behavior/Activity/ObjectNode.h"
#include "VEHA/Behavior/Activity/ActivityEdge.h"
#include "VEHA/Behavior/Activity/ActivityPartition.h"
#include "VEHA/Behavior/Activity/Activity.h"
namespace VEHA
{

ActivityNode::ActivityNode(const string& name, const string& kind) : Element(name), _kind(kind), _running(false)
{
}

ActivityNode::~ActivityNode()
{
}

vector< shared_ptr<ObjectNode> > ActivityNode::getOutgoingObjectNode(void)
{
	vector< shared_ptr<ObjectNode> > nodes;
	vector< shared_ptr<ActivityEdge> > edges = getOutgoing();
	for (size_t iE = 0; iE < edges.size() ; iE ++)
	{
		shared_ptr<ActivityNode> node = edges[iE]->getTarget();
		if (node && node->getKind() == "object")
			nodes.push_back(shared_dynamic_cast<ObjectNode>(node));

	}
	return nodes;
}

vector< shared_ptr<ObjectNode> > ActivityNode::getIncomingObjectNode(void)
{
	vector< shared_ptr<ObjectNode> > nodes;
	vector< shared_ptr<ActivityEdge> > edges = getIncoming();
	for (size_t iE = 0; iE < edges.size() ; iE ++)
	{
		shared_ptr<ActivityNode> node = edges[iE]->getSource();
		if (node && node->getKind() == "object")
			nodes.push_back(shared_dynamic_cast<ObjectNode>(node));
	}
	return nodes;
}

vector< shared_ptr<ActionNode> > ActivityNode::getOutgoingActionNode(void)
{
	vector< shared_ptr<ActionNode> > nodes;
	vector< shared_ptr<ActivityEdge> > edges = getOutgoing();
        cerr << " EDGES : " << edges.size() << endl;
	for (size_t iE = 0; iE < edges.size() ; iE ++)
	{
		shared_ptr<ActivityNode> node = edges[iE]->getTarget();
        cout << " Target : " << node->getFullName() << " kind: " << node->getKind() << endl;

        vector< shared_ptr<ActivityPartition> > toto = node->getPartition();
		if (node && node->getKind() == "action") nodes.push_back(shared_dynamic_cast<ActionNode>(node));
		else if (node && node->getKind() == "loop") nodes.push_back(shared_dynamic_cast<ActionNode>(node));
		else
		{
			vector< shared_ptr<ActionNode> > nodesTmp = node->getOutgoingActionNode();
			for (size_t iN = 0; iN < nodesTmp.size(); iN++)
			{
				nodes.push_back(nodesTmp[iN]);
			}
		}
	}
	return nodes;
}
vector<shared_ptr<ActionNode> > ActivityNode::getPossibleOutgoingActionNode()
{
	vector< shared_ptr<ActionNode> > nodes;
	vector< shared_ptr<ActivityEdge> > edges = getOutgoing();
	for (size_t iE = 0; iE < edges.size() ; iE ++)
	{
		shared_ptr<ActivityNode> node = edges[iE]->getTarget();
		if (node && node->getKind() == "action") nodes.push_back(shared_dynamic_cast<ActionNode>(node));
		else
		{
			vector< shared_ptr<ActionNode> > nodesTmp = node->getPossibleOutgoingActionNode();
			for (size_t iN = 0; iN < nodesTmp.size(); iN++)
			{
				nodes.push_back(nodesTmp[iN]);
			}
		}
	}
	return nodes;
}
vector< shared_ptr<ActionNode> > ActivityNode::getOutgoingActionNodeByPartition(const string& partitionName)
{
	vector< shared_ptr<ActionNode> > nodes;
	vector< shared_ptr<ActivityEdge> > edges = getOutgoing();
	for (size_t iE = 0; iE < edges.size() ; iE ++)
	{
		shared_ptr<ActivityNode> node = edges[iE]->getTarget();
		if (node && node->getKind() == "action")
		{
			if (node->getPartition()[0]->getName() == partitionName)
				nodes.push_back(shared_dynamic_cast<ActionNode>(node));
		}
		else
		{
			vector< shared_ptr<ActionNode> > nodesTmp = node->getOutgoingActionNode();
			for (size_t iN = 0; iN < nodesTmp.size(); iN++)
			{
				nodes.push_back(nodesTmp[iN]);
			}
		}
	}
	return nodes;
}


vector< shared_ptr<ActionNode> > ActivityNode::getIncomingActionNode(void)
{
	vector< shared_ptr<ActionNode> > nodes;
	vector< shared_ptr<ActivityEdge> > edges = getIncoming();
	for (size_t iE = 0; iE < edges.size() ; iE ++)
	{
		shared_ptr<ActivityNode> node = edges[iE]->getTarget();
		if (node && node->getKind() == "action") nodes.push_back(shared_dynamic_cast<ActionNode>(node));
		else
		{
			vector< shared_ptr<ActionNode> > nodesTmp = node->getIncomingActionNode();
			for (size_t iN = 0; iN < nodesTmp.size(); iN++)
			{
				nodes.push_back(nodesTmp[iN]);
			}
		}
	}
	return nodes;
}
vector< shared_ptr<ActionNode> > ActivityNode::getIncomingActionNodeByPartition(string partitionName)
{
	vector< shared_ptr<ActionNode> > nodes;
	vector< shared_ptr<ActivityEdge> > edges = getIncoming();
	for (size_t iE = 0; iE < edges.size() ; iE ++)
	{
		shared_ptr<ActivityNode> node = edges[iE]->getTarget();
		if (node && node->getKind() == "action")
		{
			if (node->getPartition()[0]->getName() == partitionName)
				nodes.push_back(shared_dynamic_cast<ActionNode>(node));
		}
		else
		{
			vector< shared_ptr<ActionNode> > nodesTmp = node->getIncomingActionNode();
			for (size_t iN = 0; iN < nodesTmp.size(); iN++)
			{
				nodes.push_back(nodesTmp[iN]);
			}
		}
	}
	return nodes;
}

vector< shared_ptr<ActivityEdge> > ActivityNode::getOutgoingActionNodeEdges(void)
{
	vector< shared_ptr<ActivityEdge> > actionNodeEdges;
	vector< shared_ptr<ActivityEdge> > edges = getOutgoing();
        cerr << " EDGES : " << edges.size() << endl;
	for (size_t iE = 0; iE < edges.size() ; iE ++)
	{
		shared_ptr<ActivityNode> node = edges[iE]->getTarget();
        cout << " Target : " << node->getName() << endl;

        vector< shared_ptr<ActivityPartition> > toto = node->getPartition();
		if (node && node->getKind() == "action")
		{
			actionNodeEdges.push_back(edges[iE]);
		}
		else if (node && node->getKind() == "loop")
		{
			actionNodeEdges.push_back(edges[iE]);
		}
		else
		{
			vector< shared_ptr<ActivityEdge> > nodesTmp = node->getOutgoingActionNodeEdges();
			for (size_t iN = 0; iN < nodesTmp.size(); iN++)
			{
				actionNodeEdges.push_back(nodesTmp[iN]);
			}
		}
	}
	return actionNodeEdges;
}

vector<shared_ptr<ActivityEdge> > ActivityNode::getOutgoingControlFlowEdges()
{
	vector< shared_ptr<ActivityEdge> > allOutgoing = getOutgoing();
	cerr << "Size: " << allOutgoing.size() << endl;
	vector< shared_ptr<ActivityEdge> > controlFlows;
	for(size_t i = 0; i < allOutgoing.size(); ++i)
	{
		if (allOutgoing[i]->getKind() == "ControlFlow")
			controlFlows.push_back(allOutgoing[i]);
	}
	return controlFlows;
}
}

File: Dev/CS/Mascaret/Mascaret/VEHA/Behavior/Activity/ActivityNode.cs
using System;
using System.Collections.Generic;

namespace Mascaret
{
    public class ActivityNode : Element
    {

        private string kind;
        public string Kind
        {
            get { return kind; }
            set { kind = value; }
        }


        private List<ActivityEdge> incoming = new List<ActivityEdge>();
        public List<ActivityEdge> Incoming
        {
            get { return incoming; }
            set { incoming = value; }
        }


        private List<ActivityEdge> outgoing = new List<ActivityEdge>();
        public List<ActivityEdge> Outgoing
        {
            get { return outgoing; }
            set { outgoing = value; }
        }


        private List<ActivityPartition> partitions = new List<ActivityPartition>();
        public List<ActivityPartition> Partitions
        {
            get { return partitions; }
            set { partitions = value; }
        }


        private Activity activity;
        public Activity Activity
        {
            get { return activity; }
            set { activity = value; }
        }


        private bool isRunning;
        public bool IsRunning
        {
            get { return isRunning; }
            set { isRunning = value; }
        }

        public ActivityNode(string name, string kind)
            : base(name)
        {
            this.kind = kind;
            isRunning = false;
        }

        public List<ActionNode> getIncomingActionNode()
        {
            List<ActionNode> nodes = new List<ActionNode>();
            foreach (ActivityEdge currentEdge in incoming)
            {
                if (currentEdge.Target != null && currentEdge.Target.Kind == "action")
                    nodes.Add((ActionNode)currentEdge.Target);
                else
                {
                    foreach (ActionNode currentNode in ((ActivityNode)currentEdge.Target).getIncomingActionNode())
                        nodes.Add(currentNode);
                }
            }
            return nodes;
        }

        public List<ObjectNode> getIncomingObjectNode()
        {
            List<ObjectNode> nodes = new List<ObjectNode>();
            foreach (ActivityEdge currentEdge in Incoming)
            {
                if (currentEdge.Source != null && currentEdge.Kind == "ObjectFlow")
                    nodes.Add((ObjectNode)currentEdge.Source);

            }
            return nodes;
        }

        public List<ActionNode> getIncomingActionNodeByPartition(string partitionName)
        {
            List<ActionNode> nodes = new List<ActionNode>();
            foreach (ActivityEdge currentEdge in incoming)
            {
                if (currentEdge.Target != null && currentEdge.Target.Kind == "action")
                {
                    if (currentEdge.Target.partitions != null && currentEdge.Target.partitions[0].name == partitionName)
                        nodes.Add((ActionNode)currentEdge.Target);
                }
                else
                {
                    foreach (ActionNode currentNode in currentEdge.Target.getIncomingActionNode())
                        nodes.Add(currentNode);
                }
            }
            return nodes;
        }

        public List<ActionNode> getOutgoingActionNode()
        {
            List<ActionNode> nodes = new List<ActionNode>();
            foreach (ActivityEdge currentEdge in outgoing)
            {
                //Debug.Log(" ---> " + currentEdge.Target.Kind);
                if (currentEdge.Target != null && currentEdge.Target.Kind == "action")
                {
                    nodes.Add((ActionNode)currentEdge.Target);
                    //Debug.Log(" Action : " + currentEdge.Target.getFullName());
                }
                else if (currentEdge.Target != null && currentEdge.Target.Kind == "loop")
                    nodes.Add((ActionNode)currentEdge.Target);
                else
                {
                    foreach (ActionNode currentNode in ((ActivityNode)currentEdge.Target).getOutgoingActionNode())
                        nodes.Add(currentNode);
                }
            }
            return nodes;
        }

        public List<ActionNode> getPossibleOutgoingActionNode()
        {
            List<ActionNode> nodes = new List<ActionNode>();
            foreach (ActivityEdge currentEdge in outgoing)
            {
                if (currentEdge.Target != null && currentEdge.Target.Kind == "action")
                    nodes.Add((ActionNode)currentEdge.Target);
                else
                {
                    foreach (ActionNode currentNode in currentEdge.Target.getOutgoingActionNode())
                        nodes.Add(currentNode);
                }
            }
            return nodes;
        }

        public List<ObjectNode> getOutgoingObjectNode()
        {
            List<ObjectNode> nodes = new List<ObjectNode>();
            foreach (ActivityEdge currentEdge in Outgoing)
            {
                if (currentEdge.Target != null && currentEdge.Kind == "object")
                    nodes.Add((ObjectNode)currentEdge.Target);

            }
            return nodes;
        }

        public List<ActionNode> getOutgoingActionNodeByPartition(string partitionName)
        {
            List<ActionNode> nodes = new List<ActionNode>();
            foreach (ActivityEdge currentEdge in outgoing)
            {
                if (currentEdge.Target != null && currentEdge.Target.Kind == "action")
                {
                    if (currentEdge.Target.partitions != null && currentEdge.Target.partitions[0].name == partitionName)
                        nodes.Add((ActionNode)currentEdge.Target);
                }
                else
                {
                    foreach (ActionNode currentNode in currentEdge.Target.getOutgoingActionNode())
                        nodes.Add(currentNode);
                }
            }
            return nodes;
        }

        public List<ActivityEdge> getOutgoingActionNodeEdges()
        {
            List<ActivityEdge> actionNodeEdges = new List<ActivityEdge>();
            foreach (ActivityEdge currentEdge in outgoing)
            {
                if (currentEdge.Target != null && currentEdge.Target.Kind == "action")
                    actionNodeEdges.Add(currentEdge);
                else if (currentEdge.Target != null && currentEdge.Target.Kind == "loop")
                    actionNodeEdges.Add(currentEdge);
                else
                {
                    foreach (ActivityEdge currentNode in ((ActivityNode)currentEdge.Target).getOutgoingActionNodeEdges())
                        actionNodeEdges.Add(currentNode);
                }
            }
            return actionNodeEdges;
        }

        public List<ActivityEdge> getOutgoingControlFlowEdges()
        {
            List<ActivityEdge> allOutgoing = this.Outgoing;
            List<ActivityEdge> controlFlows = new List<ActivityEdge>();
            for (int i = 0; i < allOutgoing.Count; ++i)
            {
                if (allOutgoing[i].Kind == "ControlFlow")
                    controlFlows.Add(allOutgoing[i]);
            }
            return controlFlows;
        }

    }
}
